workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs

```diff
@@ -65,27 +58,26 @@ private void GenerateComponentGroups()
                     }
 
                     typesToRemove.Add(type);
-                    addRemoveComponentActionMethod.MakeGenericMethod(type).Invoke(this, null);
                 }
-            }
+            }            
         }
 
-        private void AddRemoveComponentAction<T>()
+        private void RemoveComponents()
         {
-            var componentGroup = GetComponentGroup(ComponentType.ReadOnly<T>());
-            removeComponentActions.Add(() =>
+            var componentGroups = new List<ComponentGroup>();
+            foreach (Type type in typesToRemove)
             {
-                if (componentGroup.IsEmptyIgnoreFilter)
-                {
-                    return;
-                }
+                componentGroups.Add(GetComponentGroup(ComponentType.ReadOnly(type)));
+            }
 
+            foreach (var componentGroup in componentGroups)
+            {
                 var entityArray = componentGroup.GetEntityArray();
                 for (var i = 0; i < entityArray.Length; ++i)
                 {
-                    PostUpdateCommands.RemoveComponent<T>(entityArray[i]);
+                    EntityManager.RemoveComponent(entityArray[i], componentGroup.Types[0]);
```

Comment:
I am curious how viable is this approach?
As this is running in InternalSpatialOSCleanGroup, this system is going to be in the last group running.
Will extracting all component groups and removing them after that cause cache invalidation after all operations are completed?
Or do we need to separate removal to another system and inject components that needs removal there?

Reply:
Ah yes - this will invalidate the injected data.

Reply:
I am not very familiar with the internals - what is the good approach that will allow us not to invalidate it

Reply:
Using the `PostUpdateCommands` is the canonical way - but the method you want is inexplicably not on it. You can copy the `EntityArray` to a list and directly use the `EntityManager`